rfc/0014-metadata.md

```diff
@@ -0,0 +1,616 @@
+* Currently all the rust libraries of jsonnet seem to rely on the CPP bindings of libjsonnet. Which makes
+  extremely hard to build a statically linked binary of kwctl. There's actually an open issue for that.
+* We lose the safety of Rust types
+
+## Too much usage of annotations inside of `metadata.yml`
+
+The annotations inside of `metadata.yml` are deliberately free form. The Rust structure that
+defines `Metadata` implements them using a `HashMap<String, String>`.
+
+This means that there's no well defined Rust attribute for some of the mandatory annotations that
+`artifacthub-pkg.yml` requires.
+That makes deserialization/transformation/serialization and checks more verbose.
+
+Another possible solution would be to create a new "umbrella" configuration file that is then transformed
+into the `metadata.yml` and the `artifact-hub.yml` file.
+
+This is how the "umbrella config" might look like. This example is written using the `jsonnet` format,
+which is a superset of JSON. However we could use any other kind of format:
+
+```jsonnet
+{
+  name: 'verify-image-signatures',
+  displayName: 'Verify Image Signatures',
+  description: 'A Kubewarden Policy that verifies all the signatures of the container images referenced by a Pod',
+  keywords: [
+    'pod',
+    'signature',
+    'sigstore',
+    'trusted',
+  ],
+  authors: [
+    'Raul Cabello Martin <raul.cabello@suse.com>',
+    'Victor Cuadrado Juan <vcuadradojuan@suse.de>',
+  ],
+  homeURL: 'https://github.com/kubewarden/verify-image-signatures',
+  sourceURL: 'https://github.com/kubewarden/verify-image-signatures',
+  ociURL: 'ghcr.io/kubewarden/policies/verify-image-signatures',
+  license: 'Apache-2.0',
+  rules: [
+    {
+      apiGroups: [''],
+      apiVersions: ['v1'],
+      resources: ['pods'],
+      operations: ['CREATE', 'UPDATE'],
+    },
+  ],
+  resources: [
+    'Pod',
+    'Deployment',
+  ],
+  mutating: true,
+  contextAware: false,
+  executionMode: 'kubewarden-wapc',
+}
+```
+
+I would like to go this way honestly. I just struggle to find a good name for this umbrella file.
+
+We could then have these commands:
+
+* `kwctl scaffold artifacthub`: the command describe above, but it would read data from this umbrella config instead of `metadata.yml`
+* `kwctl scaffold metadata`: this could generate the `metadata.yml`  file
+
+Another alternative might be to not implement the `kwctl scaffold metadata`, but instead change the
+`kwctl annotate` command so that it can also read the values of the umbrella configuration file.
+
+# Alternatives
+
+## Using JSON schema
+
+We thought about using JSON Schema to describe the format of our metadata. This is interesting, it could provide
+value to our users. However it doesn't address the problem of automating the generation of
+`artifacthub-pkg.yml` and `metadata.yml`.
+
+We can revisit this topic later on.
+
+
+# Unresolved questions
+
+Should we extend the `metadata.yml` file or should we create an "umbrella" configuration file?
```

Comment:
For the policy writer, my vote goes to extending the `metadata.yml` with optional keys as specified in this rfc, instead of an umbrella jsonnet. I find extending `metadata.yml` simpler, doesn't need learning jsonnet, and provides 1 fewer level of indirection that the author needs to learn about. Can be summarized with "If you want to publish to artifacthub, add this optional fields, run kwctl scaffold, and submit there". The only hiccup is that the annotations are only a string value, but I think is worth it to deserialize twice for them or something similar.

For the implementation details, I'm still unsure if using jrsonnet & static-compiling with musl is feasible. I will have a look at it.

Reply:
Even if I think that jsonnet is cool. I don't see no big advantages to adopt now. Our needs seem to simple to worth adding another lib with the burden to learn how to maintain this new file format. In my humble opinion our needs are quite simpler. We just need to generate the artifact hub from our pre existing files. As we already have `kwctl` to do that hard work. Thus, I would say we can extend the `metadata.yml` file for now.

Reply:
However, I'm wondering if we cannot transform the `metadata.yml` file into the "umbrella" configuration file. The first idea that come to my mind after reading the question was to move the data from the annotations in the `metadata.yml` file into actual fields in the same file. Maybe using JSON format and missing the comments (no `jsonnet` at this point). With that, we have the type checks for free.  `kwctl` should be able to translate this new "umbrella" `metadata.yml` into anything (e.g. the annotation for the policy wasm and artifact hub file) with no big pain to the policy writers, isn't it? If we do that, it can be a first step to use the `jsonnet` in the future (if we decided to use it at all)

But AFAICS, that will require changes in the UI as well. Therefore, if this end to be a good idea, we may need to postpone it

Reply:
I agree with José.

For following the proposed path, I have created the outline of the cards in https://github.com/kubewarden/ui/issues/175#issuecomment-1403373903 (the current EPIC as I see it), I hope it's ok. Edits/comments/etc welcome!